EventGenerator/src/MuplusMichelGenerator_tool.cc

```diff
@@ -0,0 +1,82 @@
+#include "art/Utilities/ToolMacros.h"
+
+#include "CLHEP/Random/RandPoissonQ.h"
+#include "CLHEP/Random/RandGeneral.h"
+
+#include "Offline/EventGenerator/inc/ParticleGeneratorTool.hh"
+
+#include "Offline/DataProducts/inc/PDGCode.hh"
+#include "Offline/MCDataProducts/inc/GenId.hh"
+#include "Offline/Mu2eUtilities/inc/RandomUnitSphere.hh"
+#include "Offline/Mu2eUtilities/inc/BinnedSpectrum.hh"
+#include "Offline/GlobalConstantsService/inc/GlobalConstantsHandle.hh"
+#include "Offline/GlobalConstantsService/inc/ParticleDataList.hh"
+#include "Offline/GlobalConstantsService/inc/PhysicsParams.hh"
+
+#include "fhiclcpp/types/DelegatedParameter.h"
+
+namespace mu2e {
+  class MuplusMichelGenerator : public ParticleGeneratorTool {
+  public:
+    struct PhysConfig {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+
+      fhicl::DelegatedParameter spectrum{Name("spectrum"), Comment("Parameters for BinnedSpectrum)")};
+    };
+    typedef art::ToolConfigTable<PhysConfig> Parameters;
+
+    explicit MuplusMichelGenerator(Parameters const& conf) :
+      _pdgId(PDGCode::e_plus),
+      _mass(GlobalConstantsHandle<ParticleDataList>()->particle(_pdgId).mass()),
+      _spectrum(BinnedSpectrum(conf().spectrum.get<fhicl::ParameterSet>()))
+    {}
+
+    std::vector<ParticleGeneratorTool::Kinematic> generate() override;
+    void generate(std::unique_ptr<GenParticleCollection>& out, const IO::StoppedParticleF& stop) override;
+
+    void finishInitialization(art::RandomNumberGenerator::base_engine_t& eng, const std::string&) override {
+      _randomUnitSphere = new RandomUnitSphere(eng);
+      _randSpectrum = new CLHEP::RandGeneral(eng, _spectrum.getPDF(), _spectrum.getNbins());
+    }
+
+  private:
+    PDGCode::type _pdgId;
+    double _mass;
+
+    BinnedSpectrum    _spectrum;
+
+    RandomUnitSphere*   _randomUnitSphere;
+    CLHEP::RandGeneral* _randSpectrum;
+  };
+
+
+  std::vector<ParticleGeneratorTool::Kinematic> MuplusMichelGenerator::generate() {
+    std::vector<ParticleGeneratorTool::Kinematic>  res;
+
+    double energy = _spectrum.sample(_randSpectrum->fire());
+
+    const double p = energy * sqrt(1 - std::pow(_mass/energy,2));
+    CLHEP::Hep3Vector p3 = _randomUnitSphere->fire(p);
+    CLHEP::HepLorentzVector fourmom(p3, energy);
+
+    ParticleGeneratorTool::Kinematic k{_pdgId, ProcessCode::mu2ePrimary, fourmom};
+    res.emplace_back(k);
```

Comment:
You can write this as
res.emplace_back( _pdgId, ProcessCode::mu2ePrimary, fourmom);

That will eliminate a copy.  Maybe the optimizer is smart enough to do it anyway?  Maybe not?

The same comment 2 lines above, you can create the 3vector in the argument list to the 4 vector.

Reply:
I didn't wrote this part. It is copied from DIOGenerator_tool

Reply:
When we wrote the very first event generators emplace_back did not exist - so you were forced to pay the cost of the copy.  We are fixing these as targets of opportunity.  While no single optimization like this will really make a difference, the sum of doing everything right will.

Reply:
Thank you. That copy process was removed.